xarray/core/variable.py

```diff
@@ -893,16 +893,27 @@ def reduce(self, func, dim=None, axis=None, keep_attrs=False,
         if dim is not None and axis is not None:
             raise ValueError("cannot supply both 'axis' and 'dim' arguments")
 
+        if 'cum' in func.__name__:
+            if dim is None and axis is None:
+                raise ValueError("must supply either 'dim' or 'axis' argument to %s"
+                                 % (func.__name__))
+
         if dim is not None:
             axis = self.get_axis_num(dim)
         data = func(self.data if allow_lazy else self.values,
                     axis=axis, **kwargs)
 
         removed_axes = (range(self.ndim) if axis is None
                         else np.atleast_1d(axis) % self.ndim)
-        dims = [dim for n, dim in enumerate(self.dims)
+        dims = [adim for n, adim in enumerate(self.dims)
                 if n not in removed_axes]
 
+        def safe_shape(val):
+            return getattr(val, 'shape', ())
+
+        if safe_shape(data) == safe_shape(self.data):
```

Comment:
This should be done alternatively to calculating `dims` a few lines above, e.g.,

    if getattr(data, 'shape', ()) == self.shape:
        dims = self.dims
    else:
        removed_axes = ...
        dims = [adim for n, admin in enumerate(self.dims) ...]

Otherwise we calculate those dimensions just to throw them away